### PR TITLE
feat: Add failedToSendMessageView with to the MessageToolboxView  FS-1470

### PR DESCRIPTION
**Pull request by KaterinaWire**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/FS-1470" title="FS-1470" target="_blank">TaskFS-1470</a>  [iOS] Implement UI for sending messages when a remote backend is offline
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

Previously, we only had one type of error message with `Delete` and `Reset` buttons for error code 5xx from BE.

### Solutions

We have to distinguish between remote federation error and display different errors.

[Figma](https://www.figma.com/file/NzVYqBBPH097XIg36zEIgn/5.5.1.-Backend-offline-handling?node-id=323%3A36416&t=CvzRYF3EWLQBWKau-0)

Screenshot 2023-04-06 at 10 56 26

Screenshot 2023-04-06 at 10 56 40

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios/Wire-iOS Tests/ConversationMessageCell/MessageToolboxViewTests.swift [hunk 1]
```diff
@@ -49,7 +50,22 @@ final class MessageToolboxViewTests: CoreDataSnapshotTestCase {
...
+        message.conversationLike = otherUserConversation
+        message.failedToSendReason = .federationRemoteError
+        message.conversation?.domain = "anta.wire.link"
+    
+
+        // WHEN
+        sut.configureForMessage(message, forceShowTimestamp: true, animated: false)
+
+        // THEN
+        verifyInAllPhoneWidths(view: sut)
```
Comment (johnxnguyen) [inline wire-ios/Wire-iOS Tests/ConversationMessageCell/MessageToolboxViewTests.swift]: Is it necessary to have a snapshot for every phone width?

Comment (KaterinaWire) [reply wire-ios/Wire-iOS Tests/ConversationMessageCell/MessageToolboxViewTests.swift]: We're specifying a frame size in the setUp of this test class, and it is too small for a new error message. As far as I understand, `verifyInAllPhoneWidths` ignores this frame size and creates its own based on phone size. I can create a snapshot for one phone size, but I need to resize the frame manually.

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/Utility/MessageSendFailureView.swift [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    // MARK: Properties
+    var tapHandler: ((UIButton) -> Void)?
+
+    private let stackView = UIStackView(axis: .vertical)
+    private let titleLabel = WebLinkTextView()
+    private let retryButton: IconButton = {
+        let button = InviteButton()
+        button.titleLabel?.font = FontSpec.buttonSmallSemibold.font!
+        button.contentEdgeInsets = UIEdgeInsets(top: 8, left: 8, bottom: 8, right: 8)
+        button.setTitle(L10n.Localizable.Content.System.FailedtosendMessage.retry, for: .normal)
```
Comment (agisilaos) [inline wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/Utility/MessageSendFailureView.swift]: @KaterinaWire How about modifying our current init in InviteButton so we can pass all those properties during init? I think it makes sense. What do you think?

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/Utility/MessageSendFailureView.swift [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+        return button
+    }()
+
+    // MARK: initialization
```
Comment (agisilaos) [inline wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/Utility/MessageSendFailureView.swift]:
```suggestion
    // MARK: - initialization
```

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/Utility/MessageSendFailureView.swift [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import UIKit
+import WireCommonComponents
+import Down
+
+final class MessageSendFailureView: UIView {
+
+    // MARK: Properties
```
Comment (agisilaos) [inline wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/Utility/MessageSendFailureView.swift]:
```suggestion
    // MARK: - Properties
```

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/Utility/MessageSendFailureView.swift [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+        setupViews()
+    }
+
+    @available(*, unavailable)
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    // MARK: Setup UI
```
Comment (agisilaos) [inline wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/Utility/MessageSendFailureView.swift]:
```suggestion
    // MARK: - Setup UI
```

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/Utility/MessageSendFailureView.swift [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    // MARK: - Methods
+    @objc
+    func retryButtonTapped(_ sender: UIButton) {
+        tapHandler?(sender)
+    }
+
+}
+
```
Comment (agisilaos) [inline wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/Utility/MessageSendFailureView.swift]:
```suggestion
// MARK: - DownStyle extension
```

**Review by agisilaos — APPROVED** [review]
Left some comments on potential improvements. LGTM @KaterinaWire!

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/Utility/MessageSendFailureView.swift (hunk 2 above)
Comment (KaterinaWire) [reply wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/Utility/MessageSendFailureView.swift]: Yes, it makes sense. Please take a look

File: wire-ios/Wire-iOS Tests/ConversationMessageCell/MessageToolboxViewTests.swift (hunk 1 above)
Comment (johnxnguyen) [reply wire-ios/Wire-iOS Tests/ConversationMessageCell/MessageToolboxViewTests.swift]: It just seems a bit wasteful to have so many snapshot tests. If you want to test varying widths, perhaps it's sufficient to test a narrow and a wide width.

Comment (KaterinaWire) [reply wire-ios/Wire-iOS Tests/ConversationMessageCell/MessageToolboxViewTests.swift]: Updated snapshot tests.

**Review by johnxnguyen — APPROVED** [review]

**Pull request merged**